spring-data-relational/src/main/java/org/springframework/data/relational/repository/query/TableNameQueryPreprocessor.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright 2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.data.relational.repository.query;
+
+import org.springframework.data.relational.core.dialect.Dialect;
+import org.springframework.data.relational.core.sql.SqlIdentifier;
+import org.springframework.expression.Expression;
+import org.springframework.expression.ParserContext;
+import org.springframework.expression.spel.standard.SpelExpressionParser;
+import org.springframework.expression.spel.support.StandardEvaluationContext;
+import org.springframework.util.Assert;
+
+import java.util.regex.Pattern;
+
+public class TableNameQueryPreprocessor implements QueryPreprocessor {
+
+	private static final String EXPRESSION_PARAMETER = "$1#{";
+	private static final String QUOTED_EXPRESSION_PARAMETER = "$1__HASH__{";
+
+	private static final Pattern EXPRESSION_PARAMETER_QUOTING = Pattern.compile("([:?])#\\{");
+	private static final Pattern EXPRESSION_PARAMETER_UNQUOTING = Pattern.compile("([:?])__HASH__\\{");
+
+	private final SqlIdentifier tableName;
+	private final SqlIdentifier qualifiedTableName;
+	private final Dialect dialect;
+
+	public TableNameQueryPreprocessor(SqlIdentifier tableName, SqlIdentifier qualifiedTableName, Dialect dialect) {
```

Comment:
Introducing concepts of `QueryPreprocessing` exposes this utility quite prominently. I wonder whether we can avoid repeating table name extraction across JDBC and R2DBC modules and I wonder whether we could have a `RelationalQueryLookupStrategy`. Looking at `JdbcQueryLookupStrategy` and `R2dbcQueryLookupStrategy`, these seem to have sufficient context already for that purpose.

Reply:
Done.

I kept the `TableNameQueryPreprocessor` since putting tons of String manipulation into `RelationalQueryLookupStrategy` directly feels wrong.